utils/bids/validate.py

```diff
@@ -0,0 +1,192 @@
+    """Call command-line version of the bids validator.
+
+    Use this function if you want to parse the bids output yourself.
+    Otherwise, call validate_bids() below and it will add a description
+    of the results to the log.
+
+    Args:
+        bids_path (str): path to top directory of BIDS data.
+        out_path (pathlib path): full path and name of json formatted output
+            file that will be produced when BIDS validation is run.  If you
+            want the gear to return this file, write it to the output/
+            directory.  If you want to process this file inside the gear
+            and don't want to save it (like validate_bids() does below),
+            write it into the work/ directory.
+
+    Returns:
+
+        tuple: Two values:
+
+            * err_code (int): zero if no error.
+
+            * bids_output (dict): The results of bids validation.
+
+            `bids_output` contains a summary of the bids data present
+            and a list of errors and warnings (if any).
+    """
+
+    log.debug("Running BIDS Validator")
+
+    command = ["bids-validator", "--verbose", "--json", str(bids_path)]
+    msg = "Command: " + " ".join(command)
+    log.info(msg)
+
+    try:
+        with open(out_path, "w") as f:
+            result = sp.run(
+                command, stdout=f, stderr=sp.PIPE, universal_newlines=True, check=True
+            )
+
+    except sp.CalledProcessError as err:
+        log.error(repr(err))
+        result = sp.CompletedProcess
+        result.returncode = err.returncode
+
+    msg = command[0] + " return code: " + str(result.returncode)
+    log.info(msg)
+
+    # read validation result file to get results as dictionary
+    try:
+        with open(out_path) as jfp:
+            bids_output = json.load(jfp)
+
+    except json.JSONDecodeError as err:
+        log.error(repr(err))
+        with open(out_path) as jfp:
+            bids_output = jfp.read()
+        log.error('bids output = "%s"', bids_output)
+        result = sp.CompletedProcess
+        result.returncode = 1
+
+    return result.returncode, bids_output
+
+
+def show_errors_and_warnings(bids_output):
+    """Show what is in BIDS validation output"""
+
+    # show summary of valid BIDS stuff
+    if "summary" in bids_output:
+        msg = (
+            "bids-validator results:\n\nValid BIDS files summary:\n"
+            + pprint.pformat(bids_output["summary"], indent=8)
+            + "\n"
+        )
+        log.info(msg)
+
+    # show all errors
+    for err in bids_output["issues"]["errors"]:
+        err_msg = err["reason"] + "\n"
+        for ff in err["files"]:
+            if ff["file"]:
+                err_msg += "      In file " + ff["file"]["relativePath"]
+            if "evidence" in ff and ff["evidence"]:
+                err_msg += ", " + ff["evidence"] + "\n"
+            else:
+                err_msg += "\n"
+        log.error(err_msg)
+
+    # show all warnings
```

Comment:
Yes, that should be the procedure (validate before running the gear).  And fMRIPrep runs bids validation.  So I set the `gear-run-bids-validation` config to False.  This whole thing is probably not necessary at all.  But it is a standard part of the bids-app-template so I hesitate to take it out.  I guess it would save some space.  Hmmmm.  Belt and two pairs of suspenders.